Convert generator in Sampler back to lazy construction

ejguan opened the pull request:
Fixes #63609

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65857
* __->__ #63646

- Revert #63026 
  - Sampler is expected to be re-seeded if user specify seed before each epoch
  - Can not attach generator to self with `__iter__` because multiple iterators will ruin the use case
- Add tests to prevent the same case for different Samplers


Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

ejguan committed:
Convert Sampler back to lazily construction

[ghstack-poisoned]

ejguan committed:
Update on "Convert Sampler back to lazily construction"

Fixes #63609




[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

vfdev-5 commented on torch/utils/data/sampler.py:
```diff
@@ -89,6 +89,7 @@ def __init__(self, data_source: Sized, replacement: bool = False,
         self.replacement = replacement
         self._num_samples = num_samples
         self.generator = generator
+        self.gen: Optional[torch.Generator] = None
```
Maybe, `self.gen` can be a private attribute (no worries about later changing and BC etc) ?

ejguan replied:
Makes sense to me.
BTW, adding `gen` to self may provide a way to resume in the middle of epoch by serializing/deserializing the state of `gen`, since the `gen` is attached to the sampler instance now.

vfdev-5 reviewed (commented):
@ejguan looks good to me. Thanks. I left a nit comment.

ejguan committed:
Update on "Convert generator in Sampler back to lazy construction"

Fixes #63609


Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

VitalyFedyunin commented on torch/utils/data/sampler.py:
```diff
@@ -112,15 +113,19 @@ def num_samples(self) -> int:
         else:
-            yield from torch.randperm(n, generator=self.generator).tolist()
+            yield from torch.randperm(n, generator=self._gen).tolist()
+        self._gen = None
```
This will badly impact all currently running iterators, consider creating two individual iterators in test and yielding them in random order.

ejguan replied:
Emmm, you are right. So, we have to create iter-local generator. But, then it goes back to the problem that we can not serialize the state of this iter-local generator anymore. Then, when we resume from snapshotting, we can not really fast forward the state of generator, have to iterate over it to the certain amount of iterations.

VitalyFedyunin commented on torch/utils/data/sampler.py:
```diff
@@ -112,15 +113,19 @@ def num_samples(self) -> int:
         if self.generator is None:
-            self.generator = torch.Generator()
-            self.generator.manual_seed(int(torch.empty((), dtype=torch.int64).random_().item()))
+            if self._gen is None:
```
Wait, can you please clarify what is going on here, it looks like you want individual instance of _gen for every iterator, but placing it into `self` not going to help.

VitalyFedyunin reviewed (commented):
Not going to work as intended for two or more iterators created from same object.
```
 a = Sampler()
 i1 = iter(a)
 i2 = iter(a)
```

VitalyFedyunin commented:
Return new generator from `__iter__` to keep states shared

```
class Sampler():
    def __init__(self):
        self.iterators = {}
        pass

    def __iter__(self):
        # Everything is by pointer here
        state_dict = {}
        generator = iter(self.generator(state_dict))
        self.iterators[generator] = state_dict
        return generator

    def generator(self, state_dict):
        state_dict['gen'] = len(self.iterators)
        for i in range(10):
            yield i

s = Sampler()
i = iter(s)
i2 = iter(s)

print(i)
print(i2)

for d in i:
    print(d)

print(s.iterators)
```

ejguan commented:
> Return new generator from `__iter__` to keep states shared

Yeah, that's also something in my mind. Will update PR.

ejguan committed:
Update on "Convert generator in Sampler back to lazy construction"

Fixes #63609


Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

[ghstack-poisoned]

ejguan commented:
Have to revert it back because `torch.Generator` is still not pickable for now. Need to fix it first.

ejguan committed:
Update on "Convert generator in Sampler back to lazy construction"

Fixes #63609


Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

ejguan committed:
Update on "Convert generator in Sampler back to lazy construction"

Fixes #63609


- Revert #63026 
  - Sampler is expected to be re-seeded if user specify seed before each epoch
  - Can not attach generator to self with `__iter__` because multiple iterators will ruin the use case
- Add tests to prevent the same case for different Samplers



Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

[ghstack-poisoned]

ejguan committed:
Update on "Convert generator in Sampler back to lazy construction"

Fixes #63609


- Revert #63026 
  - Sampler is expected to be re-seeded if user specify seed before each epoch
  - Can not attach generator to self with `__iter__` because multiple iterators will ruin the use case
- Add tests to prevent the same case for different Samplers
- Keep same functionality introduced in #63026 that user can serialize the Sampler.



Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

ejguan committed:
Update on "Convert generator in Sampler back to lazy construction"

Fixes #63609


- Revert #63026 
  - Sampler is expected to be re-seeded if user specify seed before each epoch
  - Can not attach generator to self with `__iter__` because multiple iterators will ruin the use case
- Add tests to prevent the same case for different Samplers


Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

[ghstack-poisoned]

ejguan commented:
I split the original PR into two PRs.
This PR would revert Sampler back to previous state that the RNG is created lazily.

The another PR provide a new feature for potential users to serialize the iter-local generator.

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

NivekT reviewed (approved):
LGTM
